zarr/storage.py

```diff
@@ -1114,9 +1117,26 @@ def dir_path(self, path=None):
     def listdir(self, path=None):
         dir_path = self.dir_path(path)
         try:
-            out = sorted(p.rstrip('/').rsplit('/', 1)[-1]
-                         for p in self.fs.ls(dir_path, detail=False))
-            return out
+            children = sorted(p.rstrip('/').rsplit('/', 1)[-1]
+                              for p in self.fs.ls(dir_path, detail=False))
+            if self.key_separator == '/':
+                if array_meta_key in children:
+                    # special handling of directories containing an array to map nested chunk
+                    # keys back to standard chunk keys
+                    new_children = []
+                    root_path = self.dir_path(path)
+                    for entry in children:
+                        entry_path = os.path.join(root_path, entry)
+                        if _prog_number.match(entry) and self.fs.isdir(entry_path):
```

Comment:
Perhaps `ls.find` would be simpler here?

Reply:
can you elaborate? (and do you mean `list.find`)? 

Also, this code largely copied from the same method on `NestedDirectoryStore`: https://github.com/zarr-developers/zarr-python/blob/17728e86ecbcb38778cdd7b0f2e8d430a878de5a/zarr/storage.py#L1276, so if there's a cleaner way to do this then maybe I could put together a PR that cleans up `FSStore` and `NestedDirectoryStore` together.

Reply:
Sorry, I meant `fs.find`. We happen to know for fsspec implementations that this will exist, and sometimes it is more performant that walking the directory tree. `NestedDirectoryStore` probably doesn't have such a function, and walking the tree is the best you can do for local files anyway.

Reply:
using `fs.find` does clean things up a bit: see b679c08cc637e369a73625feddb42c7a956307e6